Check for Duplicates In Mining Mode

bpwhelan opened the pull request:
closes https://github.com/rampaa/JL/issues/107.

Added a config (default false) for checking duplicates when entering mining mode.

This will check all the results in mining mode for an already existing expression in the configured anki deck, and display a warning next to the result if it finds one.

Creating this as a draft as I want to:
- Take another look at the method for checking duplicates. I assume it can be simplified quite a bit... I did, however, already try creating a method that would send everything to anki in one call, since "canAddNote" can receive multiple notes, but performance was markedly worse.
- Do some extensive testing on the overall performance impact of this with large dictionary(s) as requested. With just the default dictionaries it seems to be completely unnoticeable.

rampaa commented:
> I did, however, already try creating a method that would send everything to anki in one call, since "canAddNote" can receive multiple notes, but performance was markedly worse

Interesting. Is AnkiConnect responding to, say, 100 separate `canAddNote` requests faster than a single request for those 100 notes? Or is it slower due to how we create and process those requests in JL?

rampaa commented on JL.Windows/GUI/PopupWindow.xaml.cs:
```diff
@@ -695,6 +695,31 @@ private StackPanel PrepareResultStackPanel(LookupResult result, int index, int r
+
+            if (!configManager.MineToFileInsteadOfAnki)
+            {
+                if (MiningUtils.CheckDuplicate(result).Result)
```
Accessing the `Result` property of a `Task` is a blocking operation, thus it should be avoided. Instead we should do something like

```C#
bool duplicateCard = await MiningUtils.CheckDuplicate(result).ConfigureAwait(true);
if (duplicateCard)
```

bpwhelan replied:
This is the biggest thing that I'm having trouble with since this method is synchronous, toying with something to do the check for duplicates after the window is loaded and toggling the visibility so there is 0 impact on loading the window.

rampaa commented on JL.Windows/GUI/PopupWindow.xaml.cs:
```diff
@@ -695,6 +695,31 @@ private StackPanel PrepareResultStackPanel(LookupResult result, int index, int r
 
             _ = top.Children.Add(audioButton);
+
+            if (!configManager.MineToFileInsteadOfAnki)
```
We can check whether we need to call `CheckDuplicate` at all here, like:
```C#
if (coreConfigManager.CheckForDuplicateCards
    && !configManager.MineToFileInsteadOfAnki
    && !coreConfigManager.AnkiIntegration)
```
The following check can be removed from `CheckDuplicate`:

```C#
if (!coreConfigManager.AnkiIntegration || !coreConfigManager.CheckForDuplicateCards)
{
    return false;
}
```

rampaa commented on JL.Core/Mining/MiningUtils.cs:
```diff
@@ -440,6 +440,77 @@ public static async Task MineToFile(LookupResult lookupResult, string currentTex
+        }
+
+        Dictionary<string, JLField> userFields = ankiConfig.Fields;
+        Dictionary<JLField, string> miningParams = new()
```
We probably should use `GetMiningParameters` here. The implementation provided here just partially creates mining parameters but for all we know the user might have `SourceText` as their note type's first field (which is all Anki cares for when checking whether if a card is a duplicate AFAIK) and its value is missing here.

Or better yet, we can check what's the first `Field` in the config and just create a note with that field.

rampaa commented on JL.Core/Mining/MiningUtils.cs:
```diff
@@ -440,6 +440,77 @@ public static async Task MineToFile(LookupResult lookupResult, string currentTex
+        // Audio/Picture/Video shouldn't be set here
+        // Otherwise AnkiConnect will place them under the "collection.media" folder even when it's a duplicate note
+        Note note = new(ankiConfig.DeckName, ankiConfig.ModelName, fields, ankiConfig.Tags, null, null, null, null);
+        if (!coreConfigManager.AllowDuplicateCards)
```
Why do we have this check here? As is, this method will always return false if the user enabled the `Allow duplicate cards` option, is this really the intended behavior?

bpwhelan replied:
Already removed in local.

bpwhelan commented:
IN [0d51e21](https://github.com/rampaa/JL/pull/109/commits/0d51e216a943696b852054319c75692f88952998)
- Changed this to add the icon (collapsed), and then do the duplicate check work afterwords.
- Moved around boolean logic to where we don't even enter the duplicate check method if we don't need to.
- Changed to only send one call to ankiconnect, when i got it to go successfully async, ankiconnect could not handle the requests and most of them were getting thrown out due to 503 or something.

2 things..
- I think the checkduplicates method is a bit overcomplicated... But I couldn't really find another way to do it cleanly.
- I'm not sure if there is any danger with with [this line](https://github.com/rampaa/JL/blob/0d51e216a943696b852054319c75692f88952998/JL.Windows/GUI/PopupWindow.xaml.cs#L1114). I can't get it to break, but I'm not entirely confident what would be the best way to make this safer.

bpwhelan commented:
Yeah, This is how I had it written at some point... But it was the most noticeable in terms of performance for the time I had it in.

I'm not really sure I get the technical reason for everything to be final before `UpdateLayout` but that's just a technical misunderstanding. Guess my thought was that the buttons already exist in the render, their visibility is just being toggled.

rampaa commented:
> Guess my thought was that the buttons already exist in the render, their visibility is just being toggled.

If you used `Visibility.Hidden` instead of `Visibility.Collapsed` that would mostly be correct. `Visibility.Collapsed` does not reserve space for the control in the layout, thus when you make it visible it can change the popup height/width when dynamic height/width is enabled for the popup, which can, in turn, affect where it should be positioned exactly. If we were to use `Visibility.Hidden` instead of `Visibility.Collapsed`, then we wouldn't have this problem but then we would have blank space for entries where the button is placed if it's not a duplicate.

bpwhelan commented on JL.Core/Mining/MiningUtils.cs:
```diff
@@ -440,6 +441,75 @@ public static async Task MineToFile(LookupResult lookupResult, string currentTex
+            JLField firstValue = userFields.First().Value;
+            Dictionary<JLField, string> filteredMiningParams = new()
+            {
+                { firstValue, GetMiningParameters(lookupResult, currentText, null, null, currentCharPosition, true)[firstValue] }
```
Wouldn't this mean that we would only want a select few fields to be eligible for duplicate-ness checking? This makes total sense to me, but what would that list look like? Also, i think for a few options this gets a bit messy, where we have logic above the switch, something like this, which im in the middle of writing... I'm not a big fan of this, but 
Maybe I'm misunderstanding.

```c#
    private static string GetMiningParameter(JLField field, LookupResult lookupResult, string currentText, string? formattedDefinitions, string? selectedDefinitions, int currentCharPosition, bool useHtmlTags)
    {
        string leadingSourcePart = currentText[..currentCharPosition];
        string trailingSourcePart = currentText[(currentCharPosition + lookupResult.MatchedText.Length)..];
        if (useHtmlTags)
        {
            leadingSourcePart = leadingSourcePart.ReplaceLineEndings("<br/>");
            trailingSourcePart = trailingSourcePart.ReplaceLineEndings("<br/>");
        }

        string sentence = JapaneseUtils.FindSentence(currentText, currentCharPosition);
        int searchStartIndex = currentCharPosition + lookupResult.MatchedText.Length - sentence.Length;
        if (searchStartIndex < 0 || searchStartIndex >= currentText.Length)
        {
            searchStartIndex = 0;
        }

        int sentenceStartIndex = currentText.IndexOf(sentence, searchStartIndex, StringComparison.Ordinal);
        string leadingSentencePart = currentText[sentenceStartIndex..currentCharPosition];
        string trailingSentencePart = currentText[(lookupResult.MatchedText.Length + currentCharPosition)..(sentenceStartIndex + sentence.Length)];

        return field switch
        {
            JLField.PrimarySpelling => lookupResult.PrimarySpelling,
            JLField.DictionaryName => lookupResult.Dict.Name,
            JLField.MatchedText => lookupResult.MatchedText,
            JLField.DeconjugatedMatchedText => lookupResult.DeconjugatedMatchedText,
            JLField.PrimarySpellingWithOrthographyInfo => lookupResult.PrimarySpellingOrthographyInfoList is not null
                ? $"{lookupResult.PrimarySpelling} ({string.Join(", ", lookupResult.PrimarySpellingOrthographyInfoList)})"
                : lookupResult.PrimarySpelling,
            JLField.SourceText => useHtmlTags
                ? $"{leadingSourcePart}<b>{lookupResult.MatchedText}</b>{trailingSourcePart}".ReplaceLineEndings("<br/>")
                : currentText,
            JLField.Sentence => useHtmlTags
                ? $"{leadingSentencePart}<b>{lookupResult.MatchedText}</b>{trailingSentencePart}"
                : sentence,
        };
    }
```

rampaa replied:
> Wouldn't this mean that we would only want a select few fields to be eligible for duplicate-ness checking?

No, not necessarily. I mean, we _can't_ check against audio and image fields without causing an unintended effect due to an AnkiConnect bug (see https://github.com/rampaa/JL/issues/87) but all other options should be a fair game.

> where we have logic above the switch

Sorry, I thought we could do something like the following with switch expressions:

```C#
        return field switch
        {
            JLField.Nothing or JLField.Audio or JLField.Image or _ => null,
            JLField.LeadingSourceTextPart => // The following part doesn't work
            {
                string leadingSourcePart = currentText[..currentCharPosition];
                leadingSourcePart = leadingSourcePart.ReplaceLineEndings("<br/>");
                return leadingSourcePart;
            }
        };
```
But apparently we cannot, welp. So we should use switch case instead. Something like the following ought to do it:

```C#
    private static string? GetMiningParameter(JLField field, LookupResult lookupResult, string currentText, string? formattedDefinitions, string? selectedDefinitions, int currentCharPosition)
    {
        switch (field)
        {
            case JLField.Nothing:
            case JLField.Audio:
            case JLField.Image:
            default:
                return null;

            case JLField.LeadingSentencePart:
            {
                string sentence = JapaneseUtils.FindSentence(currentText, currentCharPosition);
                int searchStartIndex = currentCharPosition + lookupResult.MatchedText.Length - sentence.Length;
                if (searchStartIndex < 0 || searchStartIndex >= currentText.Length)
                {
                    searchStartIndex = 0;
                }

                int sentenceStartIndex = currentText.IndexOf(sentence, searchStartIndex, StringComparison.Ordinal);
                return currentText[sentenceStartIndex..currentCharPosition];
            }

            case JLField.TrailingSentencePart:
            {

                string sentence = JapaneseUtils.FindSentence(currentText, currentCharPosition);
                int searchStartIndex = currentCharPosition + lookupResult.MatchedText.Length - sentence.Length;
                if (searchStartIndex < 0 || searchStartIndex >= currentText.Length)
                {
                    searchStartIndex = 0;
                }

                int sentenceStartIndex = currentText.IndexOf(sentence, searchStartIndex, StringComparison.Ordinal);
                return currentText[(lookupResult.MatchedText.Length + currentCharPosition)..(sentenceStartIndex + sentence.Length)];
            }

            case JLField.Sentence:
            {
                string sentence = JapaneseUtils.FindSentence(currentText, currentCharPosition);
                int searchStartIndex = currentCharPosition + lookupResult.MatchedText.Length - sentence.Length;
                if (searchStartIndex < 0 || searchStartIndex >= currentText.Length)
                {
                    searchStartIndex = 0;
                }

                int sentenceStartIndex = currentText.IndexOf(sentence, searchStartIndex, StringComparison.Ordinal);
                string leadingSentencePart = currentText[sentenceStartIndex..currentCharPosition];
                string trailingSentencePart = currentText[(lookupResult.MatchedText.Length + currentCharPosition)..(sentenceStartIndex + sentence.Length)];
                return $"{leadingSentencePart}<b>{lookupResult.MatchedText}</b>{trailingSentencePart}";
                ;
            }

            // Other cases
        }
    }
```

This does lead to quite a bit of code duplication but this is probably the most effective implementation we can create for our purposes.

rampaa commented on JL.Windows/GUI/PopupWindow.xaml.cs:
```diff
@@ -695,6 +699,29 @@ private StackPanel PrepareResultStackPanel(LookupResult result, int index, int r
+
+            if (!configManager.MineToFileInsteadOfAnki)
+            {
+                Button duplicate = new()
```
```suggestion
                TextBlock duplicateIconTextBlock = new()
```

This can be a `TextBlock` since we don't expect it to do anything when it's clicked.

rampaa commented on JL.Core/Mining/MiningUtils.cs:
```diff
@@ -440,6 +441,75 @@ public static async Task MineToFile(LookupResult lookupResult, string currentTex
+        Utils.Logger.Error("Sending notes to anki {notes}", JsonSerializer.Serialize(notes));
+
+        bool[]? canAddNote = await AnkiUtils.CanAddNotes(notes.ToArray()).ConfigureAwait(false);
+        if (canAddNote == null)
```
```suggestion
        if (canAddNote is null)
```

bpwhelan commented:
Everything done including suggested changes, sitting on it for a bit to test and refactor.

bpwhelan commented:
Sorry for the delay on this. in [a687c8d](https://github.com/rampaa/JL/pull/109/commits/a687c8d3d87abcb8d87cbb1dc157eac2e8699323).

- Moved Icon to the end of the row so we can actually toggle visibility of the icon instead of uncollapse it
- PR review changes like return types, array instead of List, etc.
- Changed to Textblock instead of button
- Made method to get first mining param instead of getting everything and then only getting the first.

rampaa commented on JL.Core/Mining/MiningUtils.cs:
```diff
@@ -74,6 +74,127 @@ public static class MiningUtils
 
+    private static string? GetMiningParameter(JLField field, LookupResult lookupResult, string currentText, int currentCharPosition, bool useHtmlTags)
+    {
+        string readings = "";
+        if (lookupResult.Readings is not null)
+        {
+            readings = string.Join('、', lookupResult.Readings);
+        }
```
There shouldn't be anything outside the `switch`

rampaa commented on JL.Core/Mining/MiningUtils.cs:
```diff
@@ -74,6 +74,127 @@ public static class MiningUtils
         </style>
         """;
 
+    private static string? GetMiningParameter(JLField field, LookupResult lookupResult, string currentText, int currentCharPosition, bool useHtmlTags)
```
```suggestion
    private static string? GetMiningParameter(JLField field, LookupResult lookupResult, string currentText, int currentCharPosition)
```

`useHtmlTags` will always be true for this method

rampaa commented on JL.Core/Mining/MiningUtils.cs:
```diff
@@ -74,6 +74,127 @@ public static class MiningUtils
+            case JLField.Nothing:
+            case JLField.Audio:
+            case JLField.Image:
+            case JLField.StrokeCount:
+            case JLField.KanjiGrade:
+            case JLField.RadicalNames:
+            case JLField.Definitions:
+            case JLField.SelectedDefinitions:
+            case JLField.DictionaryName:
+            case JLField.LeadingSourceTextPart:
+            case JLField.TrailingSourceTextPart:
```
Don't return null for those, return the correct values

rampaa commented on JL.Core/Mining/MiningUtils.cs:
```diff
@@ -74,6 +74,127 @@ public static class MiningUtils
+            case JLField.LeadingSourceTextPart:
+            case JLField.TrailingSourceTextPart:
+            case JLField.LocalTime:
+            case JLField.Frequencies:
+            case JLField.RawFrequencies:
+            case JLField.PreferredFrequency:
+            case JLField.FrequencyHarmonicMean:
+            case JLField.PitchAccents:
+            case JLField.NumericPitchAccents:
+            case JLField.PitchAccentForFirstReading:
+            case JLField.NumericPitchAccentForFirstReading:
+            case JLField.EdictId:
+            case JLField.DeconjugationProcess:
+            case JLField.PrimarySpellingWithOrthographyInfo:
+            case JLField.KanjiComposition:
```
Don't return null for those, return the correct values

rampaa commented on JL.Core/Mining/MiningUtils.cs:
```diff
@@ -440,6 +561,76 @@ public static async Task MineToFile(LookupResult lookupResult, string currentTex
+        {
+            LookupResult lookupResult = lookupResults[i];
+
+            Dictionary<MineType, AnkiConfig>? ankiConfigDict = await AnkiConfig.ReadAnkiConfig().ConfigureAwait(false);
+            if (ankiConfigDict is null)
+            {
+                continue;
+            }
```
```suggestion
            Dictionary<MineType, AnkiConfig>? ankiConfigDict = await AnkiConfig.ReadAnkiConfig().ConfigureAwait(false);
            if (ankiConfigDict is null)
            {
                return null;
            }
```

This should be done at the beginning of the method, not in the for loop

rampaa commented on JL.Core/Mining/MiningUtils.cs:
```diff
@@ -440,6 +561,76 @@ public static async Task MineToFile(LookupResult lookupResult, string currentTex
+            }
+
+            Dictionary<string, JLField> userFields = ankiConfig.Fields;
+            JLField firstValue = userFields.First().Value;
+            string? value = GetMiningParameter(firstValue, lookupResult, currentText, currentCharPosition, true);
+            if (value is null)
+            {
+                continue;
+            }
+
+            Dictionary<string, string> fields = ConvertFields(userFields, new()
+            {
+                [firstValue] = value
+            });
```
```suggestion
            (string firstFieldName, JLField firstField) = userFields.First();
            string? firstFieldValue = GetMiningParameter(firstField, lookupResult, currentText, currentCharPosition);
            if (value is null)
            {
                continue;
            }

            Dictionary<string, string> fields = new(1, StringComparer.Ordinal)
            {
                {  firstFieldName, "firstFieldValue" }
            };
```

bpwhelan replied:
Ah yeah this makes a lot more sense... My limited C# knowledge really showing here.

rampaa replied:
Technically speaking, calling `First()` on `userFields` isn't guaranteed to return the first field, since a `Dictionary` does not guarantee any order. However, as an implementation detail, `Dictionary` does maintain insertion order unless there are many items causing collisions, which may lead to the dictionary being rebuilt and losing the insertion order. I will properly address this after migrating JL to .NET 9 by using an `OrderedDictionary` instead of a `Dictionary` for the `Fields` property.

rampaa commented on JL.Core/Mining/MiningUtils.cs:
```diff
@@ -440,6 +561,76 @@ public static async Task MineToFile(LookupResult lookupResult, string currentTex
+                [firstValue] = value
+            });
+
+            Note note = new(ankiConfig.DeckName, ankiConfig.ModelName, fields, ankiConfig.Tags, null, null, null, null);
```
```suggestion
            Note note = new(ankiConfig.DeckName, ankiConfig.ModelName, fields, null, null, null, null, null);
```

rampaa commented on JL.Core/Mining/Anki/AnkiUtils.cs:
```diff
@@ -37,11 +37,21 @@ public static class AnkiUtils
             : null;
     }
+
+    internal static async Task<bool[]?> CanAddNotes(Note[] notes)
```
```suggestion
    internal static async ValueTask<bool[]?> CanAddNotes(List<Note> notes)
```

rampaa replied:
```suggestion
    internal static async ValueTask<List<bool>?> CanAddNotes(List<Note> notes)
```

rampaa commented on JL.Core/Mining/MiningUtils.cs:
```diff
@@ -440,6 +561,76 @@ public static async Task MineToFile(LookupResult lookupResult, string currentTex
+            positions.Add(i);
+        }
+
+        bool[]? canAddNote = await AnkiUtils.CanAddNotes(notes.ToArray()).ConfigureAwait(false);
```
```suggestion
        bool[]? canAddNote = await AnkiUtils.CanAddNotes(notes).ConfigureAwait(false);
```

rampaa commented on JL.Windows/GUI/PopupWindow.xaml.cs:
```diff
@@ -493,6 +496,13 @@ public void DisplayResults(bool generateAllResults)
             ? LastLookupResults.Count
             : Math.Min(LastLookupResults.Count, ConfigManager.Instance.MaxNumResultsNotInMiningMode);
 
+        if (MiningMode && coreConfigManager.CheckForDuplicateCards
+                    && !configManager.MineToFileInsteadOfAnki
+                    && coreConfigManager.AnkiIntegration)
+        {
+            DuplicateIcons = new(LastLookupResults.Count);
+        }
```
```suggestion
        bool checkForDuplicateCards = MiningMode
                            && coreConfigManager.CheckForDuplicateCards
                            && !configManager.MineToFileInsteadOfAnki
                            && coreConfigManager.AnkiIntegration;
                            
        TextBlock[]? duplicateIcons = checkForDuplicateCards
            ? new TextBlock[LastLookupResults.Count]
            : null;
```

rampaa commented on JL.Windows/GUI/PopupWindow.xaml.cs:
```diff
@@ -508,6 +518,7 @@ public void DisplayResults(bool generateAllResults)
         }
 
         PopupListView.ItemsSource = popupItemSource;
+        _ = CheckResultForDuplicates();
```
```suggestion
       if (checkForDuplicateCards)
       {
           _ = CheckResultForDuplicates();
       }
```

bpwhelan replied:
Seems VS Code isn't smart enough to know that duplicateIcons wouldn't be null... changing to this to avoid the warning

```
if (checkForDuplicateCards && duplicateIcons is not null)
{
    _ = CheckResultForDuplicates(duplicateIcons);
}
```

rampaa commented on JL.Windows/GUI/PopupWindow.xaml.cs:
```diff
@@ -51,6 +51,7 @@ internal sealed partial class PopupWindow
     public nint WindowHandle { get; private set; }
 
     public List<LookupResult> LastLookupResults { get; private set; } = [];
+    public List<TextBlock>? DuplicateIcons { get; private set; }
```
```suggestion
```

rampaa commented on JL.Windows/GUI/PopupWindow.xaml.cs:
```diff
@@ -793,6 +805,34 @@ private StackPanel PrepareResultStackPanel(LookupResult result, int index, int r
         }
 
+        // Keep this at the bottom
+        if (MiningMode
+            && coreConfigManager.CheckForDuplicateCards
+            && !configManager.MineToFileInsteadOfAnki
+            && coreConfigManager.AnkiIntegration
+            && DuplicateIcons is not null)
```
```suggestion
        if (duplicateIcons is not null)
```

rampaa commented on JL.Windows/GUI/PopupWindow.xaml.cs:
```diff
@@ -793,6 +805,34 @@ private StackPanel PrepareResultStackPanel(LookupResult result, int index, int r
+                Background = Brushes.Transparent,
+                Cursor = Cursors.Arrow,
+                Padding = new Thickness(0),
+                ToolTip = $"{result.PrimarySpelling} is already in the anki deck.",
```
```suggestion
                ToolTip = $"{result.PrimarySpelling} is already in the Anki deck.",
```

rampaa commented on JL.Windows/GUI/PopupWindow.xaml.cs:
```diff
@@ -793,6 +805,34 @@ private StackPanel PrepareResultStackPanel(LookupResult result, int index, int r
+            };
+
+            _ = top.Children.Add(duplicate);
+            DuplicateIcons.Add(duplicate);
+
```
```suggestion
            duplicateIcons[i] = duplicate;
```

rampaa commented on JL.Windows/GUI/PopupWindow.xaml.cs:
```diff
@@ -1067,6 +1107,31 @@ private StackPanel PrepareResultStackPanel(LookupResult result, int index, int r
 
+    private async Task CheckResultForDuplicates()
+    {
+        ConfigManager configManager = ConfigManager.Instance;
+        CoreConfigManager coreConfigManager = CoreConfigManager.Instance;
+        if (MiningMode && coreConfigManager.AnkiIntegration
+            && coreConfigManager.CheckForDuplicateCards
+            && !configManager.MineToFileInsteadOfAnki
+            && DuplicateIcons is not null)
```
```suggestion
```

rampaa commented on JL.Windows/GUI/PopupWindow.xaml.cs:
```diff
@@ -1067,6 +1107,31 @@ private StackPanel PrepareResultStackPanel(LookupResult result, int index, int r
         return stackPanel;
     }
 
+    private async Task CheckResultForDuplicates()
```
```suggestion
    private async Task CheckResultForDuplicates(TextBlock[] duplicateIcons)
```

rampaa commented on JL.Windows/GUI/PopupWindow.xaml.cs:
```diff
@@ -1067,6 +1107,31 @@ private StackPanel PrepareResultStackPanel(LookupResult result, int index, int r
+        {
+            bool[]? duplicateCard = await MiningUtils.CheckDuplicates(LastLookupResults, _currentText, _currentCharPosition).ConfigureAwait(false);
+
+            if (duplicateCard != null)
```
```suggestion
            if (duplicateCard is not null)
```

rampaa commented on JL.Windows/GUI/PopupWindow.xaml.cs:
```diff
@@ -1067,6 +1107,31 @@ private StackPanel PrepareResultStackPanel(LookupResult result, int index, int r
+                    {
+                        await MainWindow.Instance.Dispatcher.InvokeAsync(() => { DuplicateIcons[i].Visibility = Visibility.Visible; });
+                    }
+
```
```suggestion
```

rampaa commented on JL.Core/Mining/Anki/AnkiUtils.cs:
```diff
@@ -37,11 +37,21 @@ public static class AnkiUtils
+        string? resultString = response?.Result?.ToString() ?? null;
+
+        return resultString is not null
+            ? JsonSerializer.Deserialize<bool[]>(resultString)!
```
```suggestion
            ? JsonSerializer.Deserialize<List<bool>>(resultString)!
```
Deserializing it to a list instead of an array seems to be faster in this case.

bpwhelan commented:
Changes in [dbcc60a](https://github.com/rampaa/JL/pull/109/commits/dbcc60a9279a5a061c5032dfe0a3b4af99d1e9d8) all per PR. Bulk of this commit is in the `GetMiningParameter` to fill out all the ones possible, like you said just a lot of duplicate code ripped straight out of the full `GetMiningParameters` method.

- Fill out `GetMiningParameter` Method
- Move `duplicateIcons` List out of Fields
- More small Object/List Related Changes

rampaa commented on JL.Core/Mining/MiningUtils.cs:
```diff
@@ -440,6 +686,76 @@ public static async Task MineToFile(LookupResult lookupResult, string currentTex
+        List<int> positions = [];
+        bool[] results = new bool[lookupResults.Count];
+
+        Dictionary<MineType, AnkiConfig>? ankiConfigDict = await AnkiConfig.ReadAnkiConfig().ConfigureAwait(false);
+        if (ankiConfigDict is null)
+        {
+            return null;
+        }
```
This should be done at the very beginning of this method so that if we need to bail out we can do so before creating any unnecessary arrays/lists.

rampaa commented on JL.Core/Mining/MiningUtils.cs:
```diff
@@ -440,6 +686,76 @@ public static async Task MineToFile(LookupResult lookupResult, string currentTex
+            notes.Add(note);
+            positions.Add(i);
+        }
+
```
```suggestion

        if (notes.Count is 0)
        {
            return null;
        }
```

rampaa commented on JL.Core/Mining/MiningUtils.cs:
```diff
@@ -74,6 +74,252 @@ public static class MiningUtils
+                return DateTime.Now.ToString("s", CultureInfo.InvariantCulture);
+            case JLField.DictionaryName:
+                return lookupResult.Dict.Name;
+            case JLField.Frequencies:
+                if (lookupResult.Frequencies is not null)
+                {
+                    List<LookupFrequencyResult> validFrequencies = lookupResult.Frequencies
+                        .Where(static f => f.Freq is > 0 and < int.MaxValue).ToList();
+                    return LookupResultUtils.FrequenciesToText(lookupResult.Frequencies, true, lookupResult.Frequencies.Count is 1);
+                }
+                return null;
```
```suggestion
            case JLField.Frequencies:
                return lookupResult.Frequencies is not null
                    ? LookupResultUtils.FrequenciesToText(lookupResult.Frequencies, true, lookupResult.Frequencies.Count is 1)
                    : null;
```

rampaa commented on JL.Core/Mining/MiningUtils.cs:
```diff
@@ -74,6 +74,252 @@ public static class MiningUtils
+            case JLField.PreferredFrequency:
+                if (lookupResult.Frequencies is not null)
+                {
+                    List<LookupFrequencyResult> validFrequencies = lookupResult.Frequencies
+                        .Where(static f => f.Freq is > 0 and < int.MaxValue).ToList();
```
```suggestion
```

rampaa commented on JL.Core/Mining/MiningUtils.cs:
```diff
@@ -440,6 +686,76 @@ public static async Task MineToFile(LookupResult lookupResult, string currentTex
         Utils.Logger.Information("Mined {PrimarySpelling}", lookupResult.PrimarySpelling);
     }
 
+    public static async ValueTask<bool[]?> CheckDuplicates(List<LookupResult> lookupResults, string currentText, int currentCharPosition)
```
```suggestion
    public static async ValueTask<bool[]?> CheckDuplicates(LookupResult[] lookupResults, string currentText, int currentCharPosition)
```

In the master branch `LookupResults` is now an array instead of a list.

rampaa commented on JL.Core/Mining/MiningUtils.cs:
```diff
@@ -440,6 +686,76 @@ public static async Task MineToFile(LookupResult lookupResult, string currentTex
+            return null;
+        }
+
+        for (int i = 0; i < lookupResults.Count; i++)
```
```suggestion
        for (int i = 0; i < lookupResults.Length; i++)
```

rampaa commented on JL.Windows/GUI/PopupWindow.xaml.cs:
```diff
@@ -541,6 +556,7 @@ private StackPanel PrepareResultStackPanel(LookupResult result, int index, int r
         };
 
         ConfigManager configManager = ConfigManager.Instance;
+        CoreConfigManager coreConfigManager = CoreConfigManager.Instance;
```
```suggestion
```

rampaa commented on JL.Windows/GUI/PopupWindow.xaml.cs:
```diff
@@ -493,6 +495,15 @@ public void DisplayResults(bool generateAllResults)
+                            && coreConfigManager.AnkiIntegration;
+
+        TextBlock[]? duplicateIcons = checkForDuplicateCards
+            ? new TextBlock[LastLookupResults.Count]
```
```suggestion
            ? new TextBlock[LastLookupResults.Length]
```

bpwhelan commented on JL.Core/Mining/MiningUtils.cs:
```diff
@@ -74,6 +74,252 @@ public static class MiningUtils
+                string trailingSentencePart = currentText[(lookupResult.MatchedText.Length + currentCharPosition)..(sentenceStartIndex + sentence.Length)];
+                return $"{leadingSentencePart}<b>{lookupResult.MatchedText}</b>{trailingSentencePart}";
+            }
+            case JLField.SourceText:
+                string leadingSourcePart = currentText[..currentCharPosition].ReplaceLineEndings("<br/>");
+                string trailingSourcePart = currentText[(currentCharPosition + lookupResult.MatchedText.Length)..].ReplaceLineEndings("<br/>");
+                return $"{leadingSourcePart}<b>{lookupResult.MatchedText}</b>{trailingSourcePart}".ReplaceLineEndings("<br/>");
```
Yeah, I ran into that issue when writing this, looks like i forgot to add braces to that one.

rampaa commented on JL.Core/Mining/MiningUtils.cs:
```diff
@@ -74,6 +74,252 @@ public static class MiningUtils
+            case JLField.PrimarySpellingAndReadings:
+                return lookupResult.Readings is not null ? $"{lookupResult.PrimarySpelling}[{string.Join('、', lookupResult.Readings)}]" : null;
+            case JLField.PrimarySpellingAndFirstReading:
+                return $"{lookupResult.PrimarySpelling}[{(lookupResult.Readings is not null ? lookupResult.Readings[0] : "")}]";
```
```suggestion
                return lookupResult.Readings is not null
                    ? $"{lookupResult.PrimarySpelling}[{lookupResult.Readings[0]}]"
                    : null;
```

bpwhelan commented:
In [a5aee43](https://github.com/rampaa/JL/pull/109/commits/a5aee43402e7597126791f8ca67fdc9fec613570)

Add braces to case
refactor frequencies to inline conditional
get rid of unused var in Preferred Frequency
move unused and default cases to top
moved ankiconfigcheck
downmerged from upstream
changed to array instead of list per upstream change
other small pr related changes

rampaa commented:
Thanks for contributing to JL!
